### PR TITLE
Feature/iat 2095

### PR DESCRIPTION
**Pull request by jeffjohnson9046**
[IAT-2095](https://jira.fairwaytech.com/browse/IAT-2095):  Gap Report support
* Two tabs:
  * **Report Parameters:** shows the date the report was created, the filters & groups supplied by the user
  * **Report Details:** shows the details of the report
* The report data excludes `STIM` and `TUT` items; items of these types are not included in the counts
* Formulae are present; the user only has to type into the yellow "Desired" column and the spreadsheet takes care of the rest

### CODE REVIEW

File: src/main/java/org/opentestsystem/ap/imrt/iss/dto/search/Sort.java [hunk 1]
```diff
@@ -23,7 +23,7 @@
     public Sort(@JsonProperty("property") final String property,
                 @JsonProperty("direction") final Direction direction) {
         this.property = SearchProperty.findSearchProperty(Preconditions.checkNotNull(property))
-                .orElseThrow(() -> new IllegalArgumentException("Do not recognized sort property " + property));
+                .orElseThrow(() -> new IllegalArgumentException("Do not recognize sort property " + property));
```
Comment (jeffjohnson9046) [inline src/main/java/org/opentestsystem/ap/imrt/iss/dto/search/Sort.java]: Corrected minor typo here; not related to IAT-2095's feature.

File: src/main/java/org/opentestsystem/ap/imrt/iss/model/IntegerRangeSearchFilter.java [hunk 2]
```diff
@@ -26,13 +26,17 @@ public IntegerRangeSearchFilter(final SearchProperty filterProperty, final Integ
...
+    public Integer getMin() {
         return min;
     }
 
-    @VisibleForTesting
-    Integer getMax() {
+    /**
+     * @return The upper-bound of this filter
+     */
+    public Integer getMax() {
```
Comment (jeffjohnson9046) [inline src/main/java/org/opentestsystem/ap/imrt/iss/model/IntegerRangeSearchFilter.java]: Made these public so I could display their values on the **Report Parameters** sheet

File: src/main/java/org/opentestsystem/ap/imrt/iss/repository/GapReportRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                    "sum(CASE WHEN " + ITEM_DIFFICULTY_QUINTILE.getColumnName() + " = '2' THEN 1 ELSE 0 END) AS " + ACTUAL_DIFFICULTY_TWO.getFieldName() + ", \n" +
+                    "sum(CASE WHEN " + ITEM_DIFFICULTY_QUINTILE.getColumnName() + " = '3' THEN 1 ELSE 0 END) AS " + ACTUAL_DIFFICULTY_THREE.getFieldName() + ", \n" +
+                    "sum(CASE WHEN " + ITEM_DIFFICULTY_QUINTILE.getColumnName() + " = '4' THEN 1 ELSE 0 END) AS " + ACTUAL_DIFFICULTY_FOUR.getFieldName() + ", \n" +
+                    "sum(CASE WHEN " + ITEM_DIFFICULTY_QUINTILE.getColumnName() + " = '5' THEN 1 ELSE 0 END) AS " + ACTUAL_DIFFICULTY_FIVE.getFieldName() + ", \n" +
+                    "'' AS " + DELTA_DIFFICULTY_COUNT.getFieldName() + ", \n" +
+                    "'' AS " + DELTA_DIFFICULTY_ONE.getFieldName() + ", \n" +
+                    "'' AS " + DELTA_DIFFICULTY_TWO.getFieldName() + ", \n" +
+                    "'' AS " + DELTA_DIFFICULTY_THREE.getFieldName() + ", \n" +
+                    "'' AS " + DELTA_DIFFICULTY_FOUR.getFieldName() + ", \n" +
+                    "'' AS " + DELTA_DIFFICULTY_FIVE.getFieldName() + " \n";
```
Comment (jeffjohnson9046) [inline src/main/java/org/opentestsystem/ap/imrt/iss/repository/GapReportRepositoryImpl.java]: Bring back empty "placeholder" columns to make rendering the report in an Excel spreadsheet easier.

File: src/main/java/org/opentestsystem/ap/imrt/iss/model/reports/gap/GapReportBuilder.java [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+    /**
+     * Create a header for the report {@link org.apache.poi.xssf.usermodel.XSSFSheet} containing the names of the
+     * columns contained in the report.  The row will be attached to the
+     * {@link org.apache.poi.xssf.usermodel.XSSFSheet}.
+     *
+     * @param xssfSheet The {@link org.apache.poi.xssf.usermodel.XSSFSheet} that will contain the report data
+     * @param headers   A {@link java.util.Set} containin all the column header labels
+     * @param cellStyle The {@link org.apache.poi.xssf.usermodel.XSSFCellStyle} to apply to each header cell
+     */
+    private void createReportHeaderRow(final XSSFSheet xssfSheet,
```
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/iss/model/reports/gap/GapReportBuilder.java]: Can we change these void methods that alter the arguments being passed in to something like `updateReportWithHeaderRow`?  And what do you think about changing the variable name from xssfSheet here to report?  That way all the methods would indicate what is being updated.

File: src/main/java/org/opentestsystem/ap/imrt/iss/model/reports/gap/GapReportBuilder.java [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+                detailCell.setCellValue(cellContent);
+            }
+
+            // Add style specific to the "Desired" cell (i.e. the cell that requires user input).
+            if (key.toLowerCase().equals(DESIRED_DIFFICULTY_COUNT.getFieldName().toLowerCase())) {
+                desiredAndActualIndexMap.put(key, cellIndex);
+                detailCell.setCellStyle(cellStyleMap.get(CELL_STYLE_DESIRED_INPUT));
+            }
+
+            if (key.toLowerCase().startsWith("actual")) {
```
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/iss/model/reports/gap/GapReportBuilder.java]: Can you add a comment here like you did for the conditional above?  I'm not following why you're loking for the string that starts with this value.

File: src/main/java/org/opentestsystem/ap/imrt/iss/service/GapReportServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                                final GapReportBuilder gapReportBuilder) {
+        this.gapReportRepository = gapReportRepository;
+        this.gapReportBuilder = gapReportBuilder;
+    }
+
+    @Override
+    public Optional<File> getGapReport(final Collection<SearchFilter> searchFilters,
+                                       final Collection<SearchProperty> groups) {
+        if (containsInvalidFilters(searchFilters)) {
+            return Optional.empty();
```
Comment (millerts) [inline src/main/java/org/opentestsystem/ap/imrt/iss/service/GapReportServiceImpl.java]: Should this throw an IllegalArgumentException?

**Pull request merged**